sources/api/apiserver/src/server/mod.rs

```diff
@@ -646,6 +667,76 @@ fn transaction_name(query: &web::Query<HashMap<String, String>>) -> &str {
+    let mut settings_key_pair_map = HashMap::new();
+    for settings_key_pair in settings_key_pair_vec {
+        // If we see an invalid key pair. We will return the result with an error immediately.
+        let (raw_key, value) =
+            settings_key_pair
+                .split_once('=')
+                .ok_or_else(|| Error::InvalidKeyPair {
+                    input: settings_key_pair.clone(),
+                })?;
+        let mut key = Key::new(KeyType::Data, raw_key)
+            .context(error::InvalidPrefixSnafu { prefix: raw_key })?;
+        // Add "settings" prefix if the user didn't give a known prefix, to ease usage
+        let key_prefix = &key.segments()[0];
+
+        if key_prefix != "settings" {
+            let mut segments = key.segments().clone();
+            segments.insert(0, "settings".to_string());
+            key = Key::from_segments(KeyType::Data, &segments)
+                .context(error::InvalidPrefixSnafu { prefix: raw_key })?;
+        }
+
+        settings_key_pair_map.insert(key, value.to_string());
+    }
+    trace!("Key=Value map: {:#?}", settings_key_pair_map);
+    // We massage the map values to determine type of each input
+    massage_set_input(settings_key_pair_map)
+}
+
+/// We want the key=val form of 'set' to be as simple as possible; we don't want users to have to
+/// annotate or structure their input too much just to tell us the data type, but unfortunately
+/// knowledge of the data type is required to deserialize with the current datastore ser/de code.
+///
+/// To simplify usage, we use some heuristics to determine the type of each input.  We try to parse
+/// each value as a number and boolean, and if those fail, we assume a string.  (API communication
+/// is in JSON form, limiting the set of types; the API doesn't allow arrays or null, and "objects"
+/// (maps) are represented natively through our nested tree-like settings structure.)
+///
+/// If this goes wrong -- for example the user wants a string "42" -- we'll get a deserialization
+/// error, and can print a clear error and request the user use JSON input form to handle
+/// situations with more complex types.g
+///
+/// If you have an idea for how to improve deserialization so we don't have to do this, please say!
+fn massage_set_input(input_map: HashMap<Key, String>) -> Result<HashMap<Key, String>> {
+    // Deserialize the given value into the matching Rust type.  When we find a matching type, we
+    // serialize back out to the data store format, which is required to build a Settings object
+    // through the data store deserialization code.
+    let mut massaged_map = HashMap::with_capacity(input_map.len());
+    for (key, in_val) in input_map {
+        let serialized = if let Ok(b) = serde_json::from_str::<bool>(&in_val) {
+            serialize_scalar(&b).context(error::SerializeSnafu)?
+        } else if let Ok(u) = serde_json::from_str::<u64>(&in_val) {
```

Comment:
Do we have negative numbers anywhere in settings?

Edit: Oh, I guess this code was moved as-is, so I guess not!?

```suggestion
        } else if let Ok(u) = serde_json::from_str::<i64>(&in_val) {
```

Reply:
Yes, copied as is.